# Looking Through eBay Car Sales Data
The goal of this project is to skim through a csv of various car sales data using NumPy and pandas.

We will start by importing the required libraries.

In [1]:
import numpy as np
import pandas as pd

Now to load the data into a variable and take a quick peek at it.

In [2]:
autos=pd.read_csv('autos.csv', encoding='latin1')
autos.info()
autos.head()
autos.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

(50000, 20)

Taking a quick look at the info, we can see that some of the columns contain rows with varying amounts of null values, which will have to be cleaned. The column names themselves can also be renamed to be more descriptive/clear.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The columns have been copied and then manually reassigned into the .columns attribute as writing a script to rename it would have been redundant. The names have also been rewritten into snake_case to reflect a more consistent naming scheme across the project.

Next we will look some basic stats for the autos DataFrame to gain some further insight on how to clean it. Looking at the results from the .describe method, it seems that the seller and offer_type columns have the same values, so these may be ignored. The nr_of_pictures column also seems to mark every listing as having 0 pictures, so this may also be ignored. There also seem to be columns that should be numbers, however, are showing up as null values, likely due to some letters being mixed in.

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [6]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)

In [7]:
autos['price']=autos['price'].str.replace('$','')
autos['price']=autos['price'].str.replace(',','')
autos['price']=autos['price'].astype(int)
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

The odometer column also has letters mixed in with numeric data, however removing the km unit will obscure the data so we will rename the column to 'odometer_km' to make the readings easier to understand

In [8]:
autos.rename({'odometer':'odometer_km'},axis=1, inplace=True)
autos['odometer_km']=autos['odometer_km'].str.replace('km','')
autos['odometer_km']=autos['odometer_km'].str.replace(',','')
autos['odometer_km']=autos['odometer_km'].astype(int)
autos['odometer_km']

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int64

Now ww will look for outliers in the price and odometer_km columns, starting off with price:

In [9]:
print(autos['price'].unique().shape)
pd.options.display.float_format = '{:.2f}'.format
autos['price'].describe()
autos['price'].sort_values(ascending=False).head(30)

(2357,)


39705    99999999
42221    27322222
27371    12345678
47598    12345678
39377    12345678
24384    11111111
2897     11111111
11137    10000000
47634     3890000
7814      1300000
22947     1234566
514        999999
43049      999999
37585      999990
36818      350000
14715      345000
34723      299000
35923      295000
12682      265000
47337      259000
38299      250000
37840      220000
40918      198000
43668      197000
28090      194000
20351      190000
17140      180000
11433      175000
32840      169999
18509      169000
Name: price, dtype: int64

It seems that there is a big price jump from $350,000 to $999,990 so we will keep cars priced under $350,000. Next to look at the other end.

In [10]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

price
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: count, dtype: int64

We will remove the entries with a price of 0, as at no point during the listing of the car could the price be 0. There is always at least some value the seller must set the car at in order to start an auction, even if it is $1.

In [11]:
autos=autos[autos['price'].between(1,350001)]
autos.describe()

price  registration_year  power_PS  odometer_km  registration_month  \
count  48565.00           48565.00  48565.00     48565.00            48565.00   
mean    5888.94            2004.76    117.20    125770.10                5.78   
std     9059.85              88.64    200.65     39788.64                3.69   
min        1.00            1000.00      0.00      5000.00                0.00   
25%     1200.00            1999.00     71.00    125000.00                3.00   
50%     3000.00            2004.00    107.00    150000.00                6.00   
75%     7490.00            2008.00    150.00    150000.00                9.00   
max   350000.00            9999.00  17700.00    150000.00               12.00   

       postal_code  
count     48565.00  
mean      50975.75  
std       25746.97  
min        1067.00  
25%       30657.00  
50%       49716.00  
75%       71665.00  
max       99998.00

Now we will look at the odometer_km column:

In [12]:
autos['odometer_km'].describe()
autos['odometer_km'].value_counts()

odometer_km
150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: count, dtype: int64

The values in the odometer_km column seem reasonable so we will not do further cleaning.

Now we will organize the date columns:

In [13]:
autos[['date_crawled','last_seen','ad_created']][:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

The data seems to be written in YYYY-MM-DD order, lets explore each column further to see the activity on the site.

In [14]:
pd.options.display.float_format = '{:.10f}'.format
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05   0.0253268815
2016-03-06   0.0140430351
2016-03-07   0.0360135900
2016-03-08   0.0332955832
2016-03-09   0.0330896736
2016-03-10   0.0321836714
2016-03-11   0.0325748996
2016-03-12   0.0369195923
2016-03-13   0.0156697210
2016-03-14   0.0365489550
2016-03-15   0.0342839493
2016-03-16   0.0296098013
2016-03-17   0.0316277154
2016-03-18   0.0129105323
2016-03-19   0.0347781324
2016-03-20   0.0378873674
2016-03-21   0.0373725934
2016-03-22   0.0329867188
2016-03-23   0.0322248533
2016-03-24   0.0293421188
2016-03-25   0.0316071245
2016-03-26   0.0322042623
2016-03-27   0.0310923505
2016-03-28   0.0348604962
2016-03-29   0.0340986307
2016-03-30   0.0336868115
2016-03-31   0.0318336250
2016-04-01   0.0336868115
2016-04-02   0.0354782251
2016-04-03   0.0386080511
2016-04-04   0.0364871821
2016-04-05   0.0130958509
2016-04-06   0.0031710079
2016-04-07   0.0014001853
Name: proportion, dtype: float64

It seems that the listings were crawled through daily over a one month period from March to April.

In [15]:
autos['ad_created'].str[:9].value_counts(normalize=True).sort_index()

ad_created
2015-06-1   0.0000205910
2015-08-1   0.0000205910
2015-09-0   0.0000205910
2015-11-1   0.0000205910
2015-12-0   0.0000205910
2015-12-3   0.0000205910
2016-01-0   0.0000411819
2016-01-1   0.0001029548
2016-01-2   0.0001029548
2016-02-0   0.0001647277
2016-02-1   0.0002470915
2016-02-2   0.0008442294
2016-03-0   0.1419746731
2016-03-1   0.2964480593
2016-03-2   0.3336971070
2016-03-3   0.0653762998
2016-04-0   0.1608771749
Name: proportion, dtype: float64

It looks like ad creation ramped up drastically during the beginning of March.

In [16]:
autos['last_seen'].str[:10].value_counts(normalize=True).sort_index()

last_seen
2016-03-05   0.0010707299
2016-03-06   0.0043241017
2016-03-07   0.0053948317
2016-03-08   0.0074127458
2016-03-09   0.0095953876
2016-03-10   0.0106661176
2016-03-11   0.0123751673
2016-03-12   0.0237825595
2016-03-13   0.0088952950
2016-03-14   0.0126016679
2016-03-15   0.0158756306
2016-03-16   0.0164521775
2016-03-17   0.0280860702
2016-03-18   0.0073509729
2016-03-19   0.0158344487
2016-03-20   0.0206527335
2016-03-21   0.0206321425
2016-03-22   0.0213734171
2016-03-23   0.0185318645
2016-03-24   0.0197673221
2016-03-25   0.0192113662
2016-03-26   0.0168022238
2016-03-27   0.0156491300
2016-03-28   0.0208586431
2016-03-29   0.0223411922
2016-03-30   0.0247709256
2016-03-31   0.0237825595
2016-04-01   0.0227941933
2016-04-02   0.0249150623
2016-04-03   0.0252033357
2016-04-04   0.0244826521
2016-04-05   0.1247606301
2016-04-06   0.2218058272
2016-04-07   0.1319468753
Name: proportion, dtype: float64

It looks like the site was viewed daily March-April.

Lets use the .describe() method to look through the registration_year column as it is already formatted numerically:

In [17]:
autos['registration_year'].describe()

count   48565.0000000000
mean     2004.7554205704
std        88.6438870626
min      1000.0000000000
25%      1999.0000000000
50%      2004.0000000000
75%      2008.0000000000
max      9999.0000000000
Name: registration_year, dtype: float64

Taking a look at the basic stats, it looks like there are at least two invalid inputs. Let's take a closer look at the data.

In [18]:
autos['registration_year'].value_counts().sort_index(ascending=False).head(15)

registration_year
9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
Name: count, dtype: int64

In [19]:
autos['registration_year'].value_counts().sort_index(ascending=True).head(10)

registration_year
1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
Name: count, dtype: int64

Looking at the latest and earliest registration years, it looks like the earliest a car could possibly have been registered is 1910 as cars weren't invented in 1800. The latest a car could have been registered is 2016 as that is when the data was collected. Let us update the registration_year column to reflect these observations.

In [20]:
autos=autos[autos['registration_year'].between(1909,2016)]
autos['registration_year'].describe()

count   46681.0000000000
mean     2002.9107559821
std         7.1851034389
min      1910.0000000000
25%      1999.0000000000
50%      2003.0000000000
75%      2008.0000000000
max      2016.0000000000
Name: registration_year, dtype: float64

Lets look at the data by the top 10 most commonly listed brand names:

In [21]:
autos['brand'].value_counts(sort=True).head(10)

brand
volkswagen       9862
bmw              5137
opel             5022
mercedes_benz    4503
audi             4041
ford             3263
renault          2201
peugeot          1393
fiat             1197
seat              853
Name: count, dtype: int64

Lets find the mean price oof cars sold for each brand:

In [22]:
brands=autos['brand'].value_counts()
top_10_brands=brands[brands>=853].index

brand_means={}

for brand in top_10_brands:
       selected_brand=autos[autos['brand']==brand]
       brand_mean=selected_brand['price'].mean()
       brand_means[brand]=int(brand_mean)

brand_means


{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813,
 'seat': 4397}

The printed dictionary shows the top 10 most commonly sold cars and the mean price of all cars sold. The price seems to differ drastically amongst the top 10 brands.

Lets next look at mileage data to see if there is any correlation between the two:

In [23]:
mean_mileage={}

for brand in top_10_brands:
       selected_brand=autos[autos['brand']==brand]
       mileage=selected_brand['odometer_km'].mean()
       mean_mileage[brand]=int(mileage)

bmp=pd.Series(brand_means)
mm=pd.Series(mean_mileage)

bmp_mm=pd.DataFrame(bmp, columns=['brand_mean_price'])
bmp_mm['mean mileage']=mm

bmp_mm

brand_mean_price  mean mileage
volkswagen                 5402        128707
bmw                        8332        132572
opel                       2975        129310
mercedes_benz              8628        130788
audi                       9336        129157
ford                       3749        124266
renault                    2474        128071
peugeot                    3094        127153
fiat                       2813        117121
seat                       4397        121131

From the table, it seems as though there isn't a correlation between price and mileage. The odometer readings seem to all be pretty uniform. There is very slight variation as price ascends but this variation is not significant.